src/content/mods.cpp

```diff
@@ -360,51 +360,127 @@ void ModConfiguration::checkConflictsAndDeps()
-		// check which optional dependencies actually exist
-		for (const std::string &optdep : mod.optdepends) {
-			if (modnames.count(optdep) != 0)
-				mod.unsatisfied_depends.insert(optdep);
+	deps_chain.emplace(modname);
+
+	const auto &map_end = mods_by_name.end();
+
+	// Step 3: Follow the dependeny path and "push up" each
+	// mod which is required by 'modname'
+	for (const std::string &dependency : spec->depends) {
+		const auto &mod = mods_by_name.find(dependency);
+		if (mod == map_end) {
+			// Hard dependency not found
+			if (modname == spec->name)
+				spec->unsatisfied_depends.emplace(dependency);
+			return false;
 		}
-		// if a mod has no depends it is initially satisfied
-		if (mod.unsatisfied_depends.empty())
-			satisfied.push_back(mod);
-		else
-			unsatisfied.push_back(mod);
+
+		// Attempt to add "modname" to the depending mods
+		if (!mod->second.addDependingMod(modname, mods_by_name))
+			return false;
+	}
+
+	for (const std::string &dependency : spec->optdepends) {
+		const auto &mod = mods_by_name.find(dependency);
+		// It does not matter whether optional dependencies can be added,
+		// hence ignore the return value.
+		if (mod != map_end)
+			mod->second.addDependingMod(modname, mods_by_name);
+	}
+	return true;
+}
+
+void ModConfiguration::resolveDependencies()
+{
+	std::unordered_map<std::string, PrioritySortedMod> mods_by_name;
+	for (ModSpec &mod : m_unsatisfied_mods) {
+		// Reduce memory copy time by passing 'mod' as pointer
+		// 'm_unsatisfied_mods' owns the value!
+		mods_by_name.emplace(std::pair<std::string, PrioritySortedMod>(
+				mod.name, &mod));
 	}
 
-	// Step 3: mods without unmet dependencies can be appended to
-	// the sorted list.
-	while (!satisfied.empty()) {
-		ModSpec mod = satisfied.back();
-		m_sorted_mods.push_back(mod);
-		satisfied.pop_back();
-		for (auto it = unsatisfied.begin(); it != unsatisfied.end();) {
-			ModSpec &mod2 = *it;
-			mod2.unsatisfied_depends.erase(mod.name);
-			if (mod2.unsatisfied_depends.empty()) {
-				satisfied.push_back(mod2);
-				it = unsatisfied.erase(it);
+	std::vector<ModSpec> unsatisfied;
+
+	size_t old_list_size;
+	do {
+		old_list_size = mods_by_name.size();
+
+		// Step 0: Clean up old data for new weighting calculation
+		for (auto &mod : mods_by_name)
+			mod.second.deps_chain.clear();
+
+		for (auto mod_it = mods_by_name.begin(); mod_it != mods_by_name.end();) {
+			ModSpec &mod = *mod_it->second.spec;
+
+			// If the dependencies are given:
+			// Step 1: Recursively add the name to its dependency mods
+			if (mod_it->second.addDependingMod(mod.name, mods_by_name)) {
+				// Steps 3-4: See 'addDependingMod()'
+				++mod_it;
 			} else {
-				++it;
+				// One of the following cases occured:
+				// 1) Hard-dependency is missing
+				// 2) Recursive dependencies were detected
+
+				// Mod cannot be satisfied: Add to the "error list"
+				// also skip it in the next loop
+				unsatisfied.push_back(mod);
+				mods_by_name.erase(mod_it++);
```

Comment:
I think this must be `mod_it = mods_by_name.erase(mod_it);`

Reply:
There's a few places with unordered maps which use this notation, namely `m_playing_sounds` and `m_particle_spawners` across the code. I will adjust this to ease readability.

Reply:
It's not a readability thing. [`erase()`](https://en.cppreference.com/w/cpp/container/unordered_map/erase) is specified to return the element following the removed element.
Because of iterator invalidation I'm not sure if `container.erase(it); it++;` is guaranteed to work.